web/static/js/components/layout/HeaderContainer.jsx

```diff
@@ -47,42 +47,78 @@ HeaderContainer.propTypes = {
-  if (state.survey.data && state.survey.data.id == surveyId) {
-    folderId = state.survey.data.folderId
-  } else if (state.panelSurvey.data && state.panelSurvey.data.id == panelSurveyId) {
-    folderId = state.panelSurvey.data.folderId
+const getSurveyFromParams = (params, state) => {
+  const surveyId = params.surveyId && parseInt(params.surveyId)
+  if (!surveyId) return null
+  const survey = state.survey && state.survey.data
+  if (!survey) return null
+  if (survey.id == surveyId) {
+    return survey
+  } else {
+    return null
   }
-  const folder = state.folder && state.folder.folders && state.folder.folders[folderId]
-  return folder || null
 }
 
-const panelSurvey = (surveyId, state) => {
-  if (surveyId && state.survey && state.survey.data && state.survey.data.id == surveyId) {
-    const survey = state.survey.data
-    if (state.panelSurveys.items && state.panelSurveys.items[survey.panelSurveyOf]) {
-      return state.panelSurveys.items[survey.panelSurveyOf]
+const getPanelSurveyFromOccurrence = (survey, state) => {
+  if (survey) {
+    const panelSurvey = state.panelSurveys.items && state.panelSurveys.items[survey.panelSurveyId]
+    if (panelSurvey) {
+      return panelSurvey
     }
   }
   return null
 }
 
+const getPanelSurveyFromParams = (params, state) => {
+  const panelSurveyId = params.panelSurveyId && parseInt(params.panelSurveyId)
+  if (!panelSurveyId) return null
+  const panelSurvey = state.panelSurvey && state.panelSurvey.data
+  if (!panelSurvey) return null
+  if (panelSurvey.id == panelSurveyId) {
+    return panelSurvey
+  } else {
+    return null
+  }
+}
+
+const getFolderFromSurveyOrPanelSurvey = (surveyOrPanelSurvey, state) => {
+  let folderId = null
+  if (surveyOrPanelSurvey) {
+    folderId = surveyOrPanelSurvey.folderId
+  }
+  const folder = state.folder && state.folder.folders && state.folder.folders[folderId]
+  return folder || null
+}
+
 const mapStateToProps = (state, ownProps) => {
+  /**
+   * These are the survey related breadcrum cases:
+    1. Project
+    2. Project -> Survey : <Project>
+    3. Project -> Panel Survey : <Project>
+    4. Project -> Panel Survey -> Occurrence : <Project | PanelSurvey>
+    5. Project -> Folder : <Project>
+    6. Project -> Folder -> Survey : <Project | Folder>
+    7. Project -> Folder -> Panel Survey : <Project | Folder>
+    8. Project -> Folder -> Panel Survey -> Occurrence : <Project | Folder | PanelSurvey>
+  */
   const { params } = ownProps
-  const panelSurveyId = params.panelSurveyId && parseInt(params.panelSurveyId)
-  const surveyId = params.surveyId && parseInt(params.surveyId)
+  const survey = getSurveyFromParams(params, state)
+  const panelSurvey = getPanelSurveyFromParams(params, state)
+  const panelSurveyFromOccurrence = getPanelSurveyFromOccurrence(survey, state)
+  // First evaluate the panel survey from occurrence (case 8)
+  // If the survey is evaluated before (case 6), the folder is missing for the case 8.
```

Comment:
I don't finish to understand this comment 😕

Reply:
Hehe, so I need to change the comment or (worst-case) remove it 🙃 

I was trying to explain why `panelSurveyFromOccurrence || survey` works but `survey || panelSurveyFromOccurrence` doesn't.